### PR TITLE
timeline: mark as read based on the latest folded/invisible event

### PR DESCRIPTION
**Pull request by bnjbvr**
See commit by commit for an easier review.

TL;DR:

- FFI `mark_as_{un}read` has been changed to `set_unread_flag(bool)`
- a new method `Timeline::mark_as_read` is introduced in both the UI and FFI layers, attaching a read receipt to the latest event (not the event attached to the latest item)
- instead of `mark_as_unread_and_send_receipt`, it's expected of the callers to first call `set_unread_flag(false)` then `mark_as_read`
- instead of using `send_read_receipt` on the latest item's event id, when scrolled to the bottom in a room view, one would call `mark_as_read`

Pinging some app developers so they're aware of the new methods, cc @stefanceriu @ganfra @bmarty

Should fix with stuck unreads like https://github.com/element-hq/element-x-ios-rageshakes/issues/1409.

### CODE REVIEW

**Comment by ganfra** [comment]
> instead of using send_read_receipt on the latest item's event id, when scrolled to the bottom in a room view, one would call mark_as_read

We'll have to be careful when permalink will be a thing, so we don't end up marking room as read while we are stuck waiting for forward pagination.

File: bindings/matrix-sdk-ffi/src/timeline/mod.rs [hunk 1]
```diff
@@ -196,6 +196,13 @@ impl Timeline {
         })
     }
 
+    /// Mark the room as read by trying to add a read receipt to the latest
+    /// event.
+    pub async fn mark_as_read(&self, receipt_type: ReceiptType) -> Result<(), ClientError> {
+        self.inner.mark_as_read(receipt_type.into(), ReceiptThread::Unthreaded).await?;
```
Comment (stefanceriu) [inline bindings/matrix-sdk-ffi/src/timeline/mod.rs:202]: Does this still work if the latest event in the timeline is from a thread? 🤔

File: crates/matrix-sdk-ui/src/timeline/mod.rs [hunk 2]
```diff
@@ -771,6 +771,24 @@ impl Timeline {
...
+    #[instrument(skip(self), fields(room_id = ?self.room().room_id()))]
+    pub async fn mark_as_read(
+        &self,
+        receipt_type: ReceiptType,
+        thread: ReceiptThread,
+    ) -> Result<bool> {
+        if let Some(event_id) = self.inner.latest_event_id().await {
+            self.send_single_receipt(receipt_type, thread, event_id).await
+        } else {
+            trace!("can't mark room as read because there's no latest event id");
```
Comment (stefanceriu) [inline crates/matrix-sdk-ui/src/timeline/mod.rs:788]: Should this be an error instead?

**Review by stefanceriu — APPROVED** [review]
Looks sensible to me, but I agree with @ganfra, we need a better way of knowing when we're actually at the bottom/top (separately from this PR of course)

File: crates/matrix-sdk-ui/src/timeline/mod.rs (hunk 2 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:788]: Not sure. A timeline with no events would be an actual bug, right? But I think it could be happen transiently, if the timeline has been cleared because of a gappy sync, for instance.

File: bindings/matrix-sdk-ffi/src/timeline/mod.rs (hunk 1 above)
Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/timeline/mod.rs:202]: You're right, I don't think it would, *if we supported thread receipts*. But we're not, at the moment, so the unthreaded receipt should be sufficient (the thread relation doesn't matter when we're not considering them, as they appear as responses ordered by the sync order).

Comment (poljar) [reply bindings/matrix-sdk-ffi/src/timeline/mod.rs:202]: Can we document this?

File: crates/matrix-sdk-ui/src/timeline/mod.rs (hunk 2 above)
Comment (poljar) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:788]: I don't think it would be a bug? Can't we have an empty room as well?

**Review by poljar — APPROVED** [review]
Left a couple of questions, but as far as I understand this looks good. Thanks for the additional tests.

File: crates/matrix-sdk-ui/src/timeline/mod.rs (hunk 2 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:788]: I suppose that a minimal room would have some state events in the timeline for the room creation itself, and the members joining. The timeline doesn't filter them out, so they *should* be there, at least.

File: bindings/matrix-sdk-ffi/src/timeline/mod.rs (hunk 1 above)
Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/timeline/mod.rs]: I've even removed the thread parameter from `Timeline::mark_as_read`, quoting the commit message here:

It doesn't make sense to set a thread identifier for the latest event: the event might not belong to that thread, and the SDK would need to check that, since the latest event isn't reachable from the outside world (the reason why `mark_as_read` had been introduced). So let's remove it for now, and add comments related to threaded receipts.

File: crates/matrix-sdk-ui/src/timeline/mod.rs (hunk 2 above)
Comment (poljar) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:788]: I can't find the issue/comment, but turns out that an `m.room.create` event is not the first event you receive for a new room, it doesn't even have to be in the sync response that gives you the room for the first time.

That's why the `create` field in `RoomInfo` is optional :upside_down_face:.

Though ok, at least something should be there.

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:788]: TIL, thanks (also: oh wow).
